### PR TITLE
SBS add user entity

### PR DESCRIPTION
**Pull request by petro2061**
Added a new model - User. Also added user repository and user service. In user service implemented register() method for registration new user. For it added required DTO clasess. Also implemented custom annotation for check correct input repeate password.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import project.onlinebookstore.dto.user.UserRegistrationRequestDto;
+import project.onlinebookstore.dto.user.UserResponseDto;
+import project.onlinebookstore.exception.RegistrationException;
+import project.onlinebookstore.service.user.UserService;
+
+@Validated
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/dto/user/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+import project.onlinebookstore.validation.FieldMatch;
+import project.onlinebookstore.validation.Password;
+
+@Data
+@FieldMatch(first = "password", second = "repeatPassword")
+public class UserRegistrationRequestDto {
+    @Email
+    @NotBlank(message = "Field title can't be null")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/user/UserRegistrationRequestDto.java]: wrong message

File: src/main/java/project/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -33,12 +33,24 @@ public ResponseEntity<Map<String, Object>> handleValidationExceptions(
...
+        return new ResponseEntity<>(
+                getBody(ex, HttpStatus.BAD_REQUEST),
+                HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Map<String, Object>> handlerRegistrationException(
+            RegistrationException ex) {
+        return new ResponseEntity<>(
+                getBody(ex, HttpStatus.NOT_FOUND),
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
                getBody(ex, HttpStatus.CONFLICT),
```

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto registrationRequestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(registrationRequestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]: it's better to use `existsByEmail`

File: src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            throw new RegistrationException("Can't registration user. User with email: "
+                    + registrationRequestDto.getEmail()
+                    + " already exist");
+        }
+        User user = new User();
+        user.setEmail(registrationRequestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/user/impl/UserServiceImpl.java]: use mapper

File: src/main/java/project/onlinebookstore/validation/impl/FieldMatchValidator.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.firstFieldName = constraintAnnotation.first();
+        this.secondFieldName = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(Object value,
+                           ConstraintValidatorContext constraintValidatorContext) {
+        ExpressionParser expressionParser = new SpelExpressionParser();
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/validation/impl/FieldMatchValidator.java]:
```suggestion
        Object firstValue = new BeanWrapperImpl(value).getPropertyValue(this.firstFieldName);
        Object secondValue = new BeanWrapperImpl(value).getPropertyValue(this.secondFieldName);
        return Objects.equals(firstValue, secondValue);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see pr

**Comment by petro2061** [comment]
Added new implementation method isValid use BeanWrapperImpl. In reposetrory added method existsByEmail and use it in register method. Also used mapper for create user object in this method.

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**